source/extensions/access_loggers/open_telemetry/grpc_access_log_impl.h

```diff
@@ -55,7 +55,53 @@ class GrpcAccessLoggerImpl
   void initMessage() override;
   void clearMessage() override;
 
+  class OTelLogRequestCallbacks
+      : public Grpc::AsyncRequestCallbacks<
+            opentelemetry::proto::collector::logs::v1::ExportLogsServiceResponse> {
+  public:
+    OTelLogRequestCallbacks(Envoy::Event::Dispatcher& dispatcher,
+                            Common::GrpcAccessLoggerStats& stats, uint32_t sending_log_entries)
+        : dispatcher_(dispatcher), stats_(stats), sending_log_entries_(sending_log_entries) {}
+
+    void onCreateInitialMetadata(Http::RequestHeaderMap&) override {}
+
+    void onSuccess(Grpc::ResponsePtr<
+                       opentelemetry::proto::collector::logs::v1::ExportLogsServiceResponse>&& resp,
+                   Tracing::Span&) override {
+      int partial_rejected_log_entries = (resp && resp->has_partial_success())
+                                             ? resp->partial_success().rejected_log_records()
+                                             : 0;
+      stats_.logs_dropped_.add(partial_rejected_log_entries);
+      stats_.logs_written_.add(sending_log_entries_ - partial_rejected_log_entries);
+      RELEASE_ASSERT(deletion_, "`setDeletion` has not been called for initialization.");
+      deletion_();
+    }
+
+    void onFailure(Grpc::Status::GrpcStatus, const std::string&, Tracing::Span&) override {
+      stats_.logs_dropped_.add(sending_log_entries_);
+      RELEASE_ASSERT(deletion_, "`setDeletion` has not been called for initialization.");
+      deletion_();
+    }
+
+    void setDeletion(std::function<void()> deletion) { deletion_ = deletion; }
```

Comment:
If you change the deletion callback to accept OTelLogRequestCallbacks pointer and use `this` when deleting iself, then you can move deleter into constructor and avoid `RELEASE_ASSERT`.

Reply:
good point! DONE.